framework/aster-frame/src/arch/x86/tdx_guest.rs

```diff
@@ -87,3 +150,279 @@ fn handle_io(trapframe: &mut dyn TdxTrapFrame, ve_info: &tdcall::TdgVeInfo) -> b
+                }
+                Register::R11W => {
+                    trapframe.set_r11((trapframe.r11() & 0xFFFF_FFFF_FFFF_0000) | read_res)
+                }
+                Register::R12L => {
+                    trapframe.set_r12((trapframe.r12() & 0xFFFF_FFFF_FFFF_FF00) | read_res)
+                }
+                Register::R13L => {
+                    trapframe.set_r13((trapframe.r13() & 0xFFFF_FFFF_FFFF_FF00) | read_res)
+                }
+                Register::R13W => {
+                    trapframe.set_r13((trapframe.r13() & 0xFFFF_FFFF_FFFF_0000) | read_res)
+                }
+                Register::R14L => {
+                    trapframe.set_r14((trapframe.r14() & 0xFFFF_FFFF_FFFF_FF00) | read_res)
+                }
+                Register::R14D => {
+                    trapframe.set_r14((trapframe.r14() & 0xFFFF_FFFF_0000_0000) | read_res)
+                }
+                Register::R15L => {
+                    trapframe.set_r15((trapframe.r15() & 0xFFFF_FFFF_FFFF_FF00) | read_res)
+                }
+                Register::BP => {
+                    trapframe.set_rbp((trapframe.rbp() & 0xFFFF_FFFF_FFFF_0000) | read_res)
+                }
+                Register::BPL => {
+                    trapframe.set_rbp((trapframe.rbp() & 0xFFFF_FFFF_FFFF_FF00) | read_res)
+                }
+                _ => return Err(MmioError::Unimplemented),
+            }
+        },
+        Some(InstrMmioType::ReadZeroExtend) => unsafe {
+            let read_res = read_mmio(mmio_size, ve_info.guest_physical_address)
+                .map_err(|e| MmioError::TdVmcallError(e))? as usize;
+            match instr.op0_register() {
+                Register::RAX | Register::EAX | Register::AX | Register::AL => {
+                    trapframe.set_rax(read_res)
+                }
+                Register::RBX | Register::EBX | Register::BX | Register::BL => {
+                    trapframe.set_rbx(read_res)
+                }
+                Register::RCX | Register::ECX | Register::CX | Register::CL => {
+                    trapframe.set_rcx(read_res)
+                }
+                _ => return Err(MmioError::Unimplemented),
+            }
+        },
+        Some(InstrMmioType::ReadSignExtend) => return Err(MmioError::Unimplemented),
+        // MMIO was accessed with an instruction that could not be decoded or handled properly.
+        Some(InstrMmioType::Movs) => return Err(MmioError::InvalidInstruction),
+        None => return Err(MmioError::DecodeFailed),
+    }
+
+    return Ok(instr.len());
+}
+
+fn decode_instr(rip: usize) -> Result<Instruction, MmioError> {
+    if !is_kernel_vaddr(rip) {
+        return Err(MmioError::InvalidAddress);
+    }
+    let code_data = {
+        const MAX_X86_INSTR_LEN: usize = 15;
+        let mut data = [0u8; MAX_X86_INSTR_LEN];
+        unsafe {
+            core::ptr::copy_nonoverlapping(rip as *const u8, data.as_mut_ptr(), data.len());
```

Comment:
As a good programming practice, a `// SAFETY: ...` comment line should be added to explain why it is safe to use the `unsafe` here. Actually, every use of `unsafe` should have a comment to reason about the safety.